### PR TITLE
Bp 2323 buckaroo subscriptions

### PR DESCRIPTION
**Pull request by vegimcarkaxhija**

### CODE REVIEW

File: gateway-buckaroo-sofort.php [hunk 1]
```diff
@@ -8,18 +8,19 @@
...
         $this->method_title           = "Buckaroo Sofort";
         $this->setIcon('24x24/sofort.png', 'svg/sofort.svg');
-        
-        $this->migrateOldSettings('woocommerce_buckaroo_sofortbanking_settings');
+
         parent::__construct();
+        $this->migrateOldSettings('woocommerce_buckaroo_sofortbanking_settings');
         $this->addRefundSupport();
-
+        $this->addSubscriptionsSupport();
```
Comment (harli91) [inline gateway-buckaroo-sofort.php:23]: as I sad before you should use hooks/filters so this should be `apply_filter('buckaroo_init_payment_class', $this)` then in your plugin you add  support for subscription

File: gateway-buckaroo-sofort.php [hunk 2]
```diff
@@ -45,9 +46,13 @@ public function process_payment($order_id)
         $order = getWCOrder($order_id);
         /** @var BuckarooSofortbanking */
         $sofortbanking = $this->createDebitRequest($order);
-        $response = $sofortbanking->Pay();
-
-        return fn_buckaroo_process_response($this, $response);
+        if($this->is_subscriptions_enabled() && $this->has_subscription($order_id)){
```
Comment (harli91) [inline gateway-buckaroo-sofort.php:49]: same here you can do this checking and order in the subscription plugin you just do 
```
$sofortbanking = $this->createDebitRequest($order);
apply_filters('buckaroo_before_payment_request', $order, $sofortbanking);
$response = $sofortbanking->Pay();
return fn_buckaroo_process_response($this, $response);
```

**Review by harli91 — CHANGES_REQUESTED** [review]

**Review by harli91 — APPROVED** [review]

File: gateway-buckaroo.php [hunk 3]
```diff
@@ -1390,4 +1390,21 @@ protected function create_capture_error($message)
...
+     * @return array
+     */
+    function apply_filters_or_error($tag, $value, ...$args) {
+        if (has_filter($tag)) {
+            return apply_filters($tag, $value, ...$args);
+        } else {
+            // Handle the case when the filter is not found
+            // You can return an error message, throw an exception, or take any other action as needed
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (vegimcarkaxhija) [reply gateway-buckaroo.php]: done

File: gateway-buckaroo.php [hunk 4]
```diff
@@ -1400,9 +1400,11 @@ protected function create_capture_error($message)
...
     function apply_filters_or_error($tag, $value, ...$args) {
-        if (has_filter($tag)) {
-            return apply_filters($tag, $value, ...$args);
+        if (!has_filter($tag)) {
+            return null;
         }
-        return null;
+        $response = apply_filters($tag, $value, ...$args);
+
+        return (isset($response['result']) && $response['result'] === 'no_subscription') ? null : $response;
```
Comment (harli91) [inline gateway-buckaroo.php:1408]: why not return the response directly `apply_filters($tag, $value, ...$args);` and in the subscription check we don't return a array but `null`?

**Review by harli91 — APPROVED** [review]

**Review by harli91 — APPROVED** [review]

**Pull request merged**